rosidl_generator_mypy/__init__.py

```diff
@@ -102,33 +109,55 @@ def get_defined_classes(content: IdlContent) -> Set[str]:
 
 def to_type_annotation(
     current_namespace: NamespacedType, defined_classes: Set[str], type_: AbstractType
-) -> str:
+) -> Annotation:
     if isinstance(type_, NamespacedType):
         if type_.namespaces == current_namespace.namespaces:
             if type_.name in defined_classes:
                 # member is defined in the same module, so no need to add namespaces
-                return '"{}"'.format(type_.name)
+                annotation = '"{}"'.format(type_.name)
+                return Annotation(annotation, annotation)
 
             # NOTE: We export .pyi files, which don't affect the Python code execution at all.
             # As mypy solves the import cycles properly,
             # we import classes from not a module but a package.
             # (i.e. in the same way as imports for other packages)
 
-        return "{}.{}".format(".".join(type_.namespaces), type_.name)
+        annotation = "{}.{}".format(".".join(type_.namespaces), type_.name)
+        return Annotation(annotation, annotation)
 
     try:
         ret = generate_py_impl.get_python_type(type_)
         if ret is not None:
-            return str(ret)
+            return Annotation(str(ret), str(ret))
     except Exception:
         pass
 
-    if isinstance(type_, (AbstractSequence, Array)):
-        return "typing.Sequence[{}]".format(
-            to_type_annotation(current_namespace, defined_classes, type_.value_type)
+    if isinstance(type_, Array):
+        type_annotation = to_type_annotation(
+            current_namespace, defined_classes, type_.value_type
+        )
+        return Annotation(
```

Comment:
I read the specification.
https://design.ros2.org/articles/idl_interface_definition.html

According to the `Type Mapping` section on the document, shouldn't we check `type_annotation.getter in SPECIAL_NESTED_BASIC_TYPES` in the same way as `AbstractSequence`?

Reply:
I'm not sure... I've tested this against most of the standard ros interfaces (eg sensor_msgs) and I _think_ everything looked correct. 
It could be prudent to write some tests which `colcon build` a bunch of messages and make assertions about the `pyi` code it generates but I'm not sure I'll have time to do that right now.

Reply:
I referred to the `rosidl_python` implementation, and it actually follows the specification.

I think the member initialization part is helpful to understand this behavior:
https://github.com/ros2/rosidl_python/blob/6675823dd6123360912e233a830d1edab89c7651/rosidl_generator_py/resource/_msg.py.em#L320-L332

I've also tested with the following message definition and got the same result that I expected.

### Definition
```
int64[] integer_array
int64[4] integer_fixed_array

std_msgs/Header[] header_array
std_msgs/Header[] header_fixed_array
```

### integer_array

```py
    @integer_array.setter
    def integer_array(self, value):
        if isinstance(value, array.array):
            assert value.typecode == 'q', \
                "The 'integer_array' array.array() must have the type code of 'q'"
            self._integer_array = value
            return
        if __debug__:
            from collections.abc import Sequence
            from collections.abc import Set
            from collections import UserList
            from collections import UserString
            assert \
                ((isinstance(value, Sequence) or
                  isinstance(value, Set) or
                  isinstance(value, UserList)) and
                 not isinstance(value, str) and
                 not isinstance(value, UserString) and
                 all(isinstance(v, int) for v in value) and
                 all(val >= -9223372036854775808 and val < 9223372036854775808 for val in value)), \
                "The 'integer_array' field must be a set or sequence and each value of type 'int' and each integer in [-9223372036854775808, 9223372036854775807]"
        self._integer_array = array.array('q', value)
```

### integer_fixed_array

```py
    @integer_fixed_array.setter
    def integer_fixed_array(self, value):
        if isinstance(value, numpy.ndarray):
            assert value.dtype == numpy.int64, \
                "The 'integer_fixed_array' numpy.ndarray() must have the dtype of 'numpy.int64'"
            assert value.size == 4, \
                "The 'integer_fixed_array' numpy.ndarray() must have a size of 4"
            self._integer_fixed_array = value
            return
        if __debug__:
            from collections.abc import Sequence
            from collections.abc import Set
            from collections import UserList
            from collections import UserString
            assert \
                ((isinstance(value, Sequence) or
                  isinstance(value, Set) or
                  isinstance(value, UserList)) and
                 not isinstance(value, str) and
                 not isinstance(value, UserString) and
                 len(value) == 4 and
                 all(isinstance(v, int) for v in value) and
                 all(val >= -9223372036854775808 and val < 9223372036854775808 for val in value)), \
                "The 'integer_fixed_array' field must be a set or sequence with length 4 and each value of type 'int' and each integer in [-9223372036854775808, 9223372036854775807]"
        self._integer_fixed_array = numpy.array(value, dtype=numpy.int64)
```

### header_array

```py
    @header_array.setter
    def header_array(self, value):
        if __debug__:
            from std_msgs.msg import Header
            from collections.abc import Sequence
            from collections.abc import Set
            from collections import UserList
            from collections import UserString
            assert \
                ((isinstance(value, Sequence) or
                  isinstance(value, Set) or
                  isinstance(value, UserList)) and
                 not isinstance(value, str) and
                 not isinstance(value, UserString) and
                 all(isinstance(v, Header) for v in value) and
                 True), \
                "The 'header_array' field must be a set or sequence and each value of type 'Header'"
        self._header_array = value
```

### header_fixed_array

```py
    @header_fixed_array.setter
    def header_fixed_array(self, value):
        if __debug__:
            from std_msgs.msg import Header
            from collections.abc import Sequence
            from collections.abc import Set
            from collections import UserList
            from collections import UserString
            assert \
                ((isinstance(value, Sequence) or
                  isinstance(value, Set) or
                  isinstance(value, UserList)) and
                 not isinstance(value, str) and
                 not isinstance(value, UserString) and
                 len(value) == 4 and
                 all(isinstance(v, Header) for v in value) and
                 True), \
                "The 'header_fixed_array' field must be a set or sequence with length 4 and each value of type 'Header'"
        self._header_fixed_array = value
```